// smooch-core/lib/smooch.js
'use strict';

exports.__esModule = true;
exports.Smooch = exports.SERVICE_URL = undefined;

var _auth = require('./utils/auth');

var _appUsers = require('./api/appUsers');

var _conversations = require('./api/conversations');

var _stripe = require('./api/stripe');

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var SERVICE_URL = exports.SERVICE_URL = 'https://api.smooch.io/v1';

var Smooch = exports.Smooch = function Smooch() {
    var auth = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, Smooch);

    var _options$serviceUrl = options.serviceUrl;
    var serviceUrl = _options$serviceUrl === undefined ? SERVICE_URL : _options$serviceUrl;
    var _options$headers = options.headers;
    var headers = _options$headers === undefined ? {} : _options$headers;

    this.VERSION = _package2.default.version;
    this.serviceUrl = serviceUrl;

    if (auth.keyId || auth.secret) {
        throw new Error('Key Id or Secret should not be used on the browser side. You must generate a JWT beforehand.');
    }

    this.headers = headers;
    this.authHeaders = (0, _auth.getAuthenticationHeaders)(auth);

    this.appUsers = new _appUsers.AppUsersApi(this.serviceUrl, this.authHeaders, this.headers);
    this.conversations = new _conversations.ConversationsApi(this.serviceUrl, this.authHeaders, this.headers);
    this.stripe = new _stripe.StripeApi(this.serviceUrl, this.authHeaders, this.headers);

    this.utils = {};
};
