// smooch-core/lib/api/webhooks.js
'use strict';

exports.__esModule = true;
exports.WebhooksApi = undefined;

var _base = require('./base');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Webhook properties
 * @typedef WebhookProps
 */

/**
 * @class WebhooksApi
 * @extends BaseApi
 */

var WebhooksApi = exports.WebhooksApi = function (_BaseApi) {
    _inherits(WebhooksApi, _BaseApi);

    function WebhooksApi() {
        _classCallCheck(this, WebhooksApi);

        var _this = _possibleConstructorReturn(this, _BaseApi.apply(this, arguments));

        _this.allowedAuth = ['jwt'];
        return _this;
    }

    /**
     * Validates the properties sent to the API
     * @param  {WebhookProps}  props      - a properties object
     * @param  {Boolean} isTargetRequired - tells if the target property is required (i.e., on creation) [default = false]
     * @return {WebhookProps}             - the properties object passed in parameter
     */


    WebhooksApi.prototype.validateProps = function validateProps(props) {
        var isTargetRequired = arguments.length <= 1 || arguments[1] === undefined ? false : arguments[1];

        if (!props || Object.keys(props).length === 0) {
            return Promise.reject(new Error('Must provide props.'));
        }

        if (isTargetRequired && !props.target) {
            return Promise.reject(new Error('Must provide a target.'));
        }

        if (props.target && !props.target.startsWith('http://') && !props.target.startsWith('https://')) {
            return Promise.reject(new Error('Malformed target url.'));
        }

        return Promise.resolve(props);
    };

    /**
     * List all webhooks
     * @param  {AuthCredentials} auth
     * @return {APIResponse}
     */


    WebhooksApi.prototype.list = function list() {
        var url = this.getFullURL('webhooks');
        return this.request('GET', url);
    };

    /**
     * Creates a webhook
     * @param  {WebhookProps} props - a properties object
     * @return {APIResponse}
     */


    WebhooksApi.prototype.create = function create(props) {
        var _this2 = this;

        var url = this.getFullURL('webhooks');
        return this.validateProps(props, true).then(function (validatedProps) {
            return _this2.request('POST', url, validatedProps);
        });
    };

    /**
     * Retrieves a webhook
     * @param  {string} webhookId - an id
     * @return {APIResponse}
     */


    WebhooksApi.prototype.get = function get(webhookId) {
        var url = this.getFullURL('webhooks', webhookId);
        return this.request('GET', url);
    };

    /**
     * Updates a webhook
     * @param  {string} webhookId    - an id
     * @param  {WebhookProps} props  - a properties object
     * @return {APIResponse}
     */


    WebhooksApi.prototype.update = function update(webhookId, props) {
        var _this3 = this;

        var url = this.getFullURL('webhooks', webhookId);
        return this.validateProps(props).then(function (validatedProps) {
            return _this3.request('PUT', url, validatedProps);
        });
    };

    /**
     * Deletes a webhook
     * @param  {string} webhookId - an id
     * @return {APIResponse}
     import { getAuthenticationHeaders } from '../../../src/utils/auth';
     */


    WebhooksApi.prototype.delete = function _delete(webhookId) {
        var url = this.getFullURL('webhooks', webhookId);
        return this.request('DELETE', url);
    };

    return WebhooksApi;
}(_base.BaseApi);
